bootloader/piboot.go

```diff
@@ -0,0 +1,441 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package bootloader
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader/pibootenv"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+const (
+	pibootCfgFilename = "piboot.conf"
+	pibootPartFolder  = "/piboot/ubuntu/"
+	// TODO Use boot.InitramfsUbuntuSeedDir, boot.InitramfsUbuntuBootDir
+	// Need to solve circular dependency
+	runMntDir     = "/run/mnt/"
+	ubuntuSeedDir = runMntDir + "ubuntu-seed/"
+	ubuntuBootDir = runMntDir + "ubuntu-boot/"
+)
+
+type piboot struct {
+	rootdir          string
+	basedir          string
+	prepareImageTime bool
+}
+
+func (p *piboot) setDefaults() {
+	p.basedir = "/boot/piboot/"
+}
+
+func (p *piboot) processBlOpts(blOpts *Options) {
+	if blOpts == nil {
+		return
+	}
+
+	p.prepareImageTime = blOpts.PrepareImageTime
+	switch {
+	case blOpts.Role == RoleRecovery || blOpts.NoSlashBoot:
+		// RoleRecovery or NoSlashBoot imply we use
+		// the environment file in /piboot/ubuntu as
+		// it exists on the partition directly
+		p.basedir = pibootPartFolder
+	}
+}
+
+// newPiboot creates a new Piboot bootloader object
+func newPiboot(rootdir string, blOpts *Options) Bootloader {
+	p := &piboot{
+		rootdir: rootdir,
+	}
+	p.setDefaults()
+	p.processBlOpts(blOpts)
+	return p
+}
+
+func (p *piboot) Name() string {
+	return "piboot"
+}
+
+func (p *piboot) dir() string {
+	if p.rootdir == "" {
+		panic("internal error: unset rootdir")
+	}
+	return filepath.Join(p.rootdir, p.basedir)
+}
+
+func (p *piboot) envFile() string {
+	return filepath.Join(p.dir(), pibootCfgFilename)
+}
+
+// piboot enabled if env file exists
+func (p *piboot) Present() (bool, error) {
+	return osutil.FileExists(p.envFile()), nil
+}
+
+// Variables stored in ubuntu-seed:
+//   snapd_recovery_system
+//   snapd_recovery_mode
+//   snapd_recovery_kernel
+// Variables stored in ubuntu-boot:
+//   kernel_status
+//   snap_kernel
+//   snap_try_kernel
+//   snapd_extra_cmdline_args
+//   snapd_full_cmdline_args
+//   recovery_system_status
+//   try_recovery_system
+func (p *piboot) SetBootVars(values map[string]string) error {
+	env := pibootenv.NewEnv(p.envFile())
```

Comment:
if this bootenv file is only going to be read by snapd and never by any actual bootscripts, could we use uboot.env files instead which are error correcting instead of a basic key=value format file?

Reply:
Yeah, I think we can do this, some extra checks are always good.